[docs] Refresh Developer Policy text

rnk opened the pull request:
Clarify lots of existing practice. Expand on the "major change" section,
which is the closest thing we have on how to run an RFC.

rnk committed:
[docs] Refresh Developer Policy text

Clarify lots of existing practice. Expand on the "major change" section,
which is the closest thing we have on how to run an RFC.

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
-#. Once you have created your patch, create a
-   :ref:`GitHub Pull Request <github-reviews>` for
-   it (or commit it directly if applicable).
+* Identify 2-3 individuals to review the patch. Look through the relevant
```
Do random GIthub users have permissions to add PR reviewers? If not, we should document the recommended way of attracting attention, e.g., by @-pinging the relevant people in a comment and explicitly stating you don't have permission to add reviewers (for one-off PRs) and/or asking for "triage" set of permissions.

AaronBallman replied:
> Do random GIthub users have permissions to add PR reviewers?

In my experience, they do not have the ability to do this themselves unless they're a member of the organization. Though I wonder if that can be relaxed?

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
+
+* Make sure your patch is based on a recent commit from ``main``, rather than a
+  previous release branch. If you want to make changes to a release branch, land
+  a change in ``main`` first and ask a release manager to backport it.
```
Should we encourage developers to backport their patches themselves, when possible?

rnk replied:
Yes, done, sorry I missed that.

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
+archived, and that notices of confidentiality or non-disclosure cannot be
+respected.
+
+We accept code contributions as :ref:`GitHub Pull Requests <github-reviews>`.
```
Some thing I don't have a strong opinion on, but witnessed: some PRs or Github issue discussions grow very long and sometimes contentious, but much fewer people engage with them. Maybe we want some sort of soft guideline on when to move such a discussion to the forums for broader community involvement.

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -220,27 +218,34 @@ Test Cases
-  by :doc:`bugpoint <Bugpoint>` or manually. It is unacceptable to place an
-  entire failing program into ``llvm/test`` as this creates a *time-to-test*
-  burden on all developers. Please keep them short.
+* Test cases are written in the relevant input language of the relevant
```
Maybe we want to highlight more prominently that, whenever possible, tests should be written against a tool such as `opt`rather than as unit tests and use command-line tools such as `FileCheck` to verify the output for testing efficiency reasons.

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -336,12 +341,12 @@ you follow these guidelines to help review, search in logs, email formatting
-set right" will leave the reviewer wondering about which bits, and why they
-weren't right, while "Correctly set overflow bits in TargetInfo" conveys almost
-all there is to the change.
+Commit messages should communicate briefly what the change does, and provide
```
IMO, the "why" part is the most important in the commit message. We can understand the "what" from the commit itself, although it is useful to summarize the "what" in the message for the reviewer to check whether the code actually does what the author intended.

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -336,12 +341,12 @@ you follow these guidelines to help review, search in logs, email formatting
+messages should be thoughtfully written and specific, rather than vague. For
+example, "bits were not set right" will leave the reviewer wondering about which
+bits, and why they weren't right, while "Correctly set overflow bits in
+TargetInfo" conveys almost all there is to the change.
```
I'd also remind that the final commit, potentially squashed from multiple commits in a PR, will have the description taken from the PR description. The latter is set once when the PR is created, defaulting to the commit message if the PR only contains one commit, and does _not_ track the changes to the commit message. As a corollary, developers must manually update the PR description as they iterate on the change...

ftynse commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -552,26 +556,42 @@ to do so.
+often be helpful in making design discussions more concrete by demonstrating
+what is possible.
+
+TODO: Elaborate on best practices for making a successful RFC.
```
Random ideas for this from recent experience:

 - make it targeted, don't touch components irrelevant to the task
 - explain how the change improves LLVM for all/most stakeholders rather than your specific use case
 - periodically summarize the current state of the discussion and clearly separate points where consensus seems to emerge from those where further discussion is necessary
 - by default, the burden of proof is on the proposer (though we'd like to avoid situations when somebody keeps asking for more proof as a way of exhausting the proposer)

rnk replied:
I added that list, but I'm sure we can add more here :) I think my goal with this PR was to get a better base for future edits.

AaronBallman commented:
> which is the closest thing we have on how to run an RFC.

FYI:
https://clang.llvm.org/get_involved.html#criteria, starting from `The Clang community uses an RFC process `

AaronBallman commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
+distributed open source project. It enables long-form asyncronous text
+communication, and this is where people tend to propose major changes or
+propose new designs in the form of RFCs (Request For Comment), which are
+described later. Please be aware that the discourse forums are public and
```
```suggestion
described later. Please be aware that the Discourse forums are public and
```

AaronBallman commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
 
-When making a patch for review, the goal is to make it as easy for the reviewer
-to read it as possible.  As such, we recommend that you:
+Submitting patches for review is straightforward with GitHub. Follow the
```
> Submitting patches for review is straightforward with GitHub.

I'd probably reword this to "Patches are submitted through GitHub" rather than claim it's straightforward for everyone (not everyone is familiar with GitHub or comfortable with git in general).

AaronBallman commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
+  :ref:`Maintainers` file or browse git blame for likely stakeholders for the
+  code want to modify.
+
+* Make sure your patch is based on a recent commit from ``main``, rather than a
+  previous release branch. If you want to make changes to a release branch, land
```
```suggestion
* To avoid precommit CI failures due to merge conflicts, base your patches on a recent commit from ``main``. If you want to make changes to a release branch, land
```
Maybe? It explains why this is needed.

AaronBallman commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -220,27 +218,34 @@ Test Cases
+feature tests only. More extensive test cases (e.g., entire applications,
+benchmarks, etc) should be added to the ``llvm-test`` test suite.  The ``llvm-test``
+suite is for integration and application testing (correctness, performance, etc)
+testing, not feature or regression testing.
```
Do we also want to mention that it's the place to put tests which are incompatible with the license from the main LLVM repo?

AaronBallman reviewed (commented):
Thank you for working on this refresh, it's definitely heading in a good direction IMO! I have some light feedback, still thinking about content further.

rnk committed:
Merge branch 'main' into dev-policy-refresh

rnk committed:
fix grammar nit

Co-authored-by: Oleksandr "Alex" Zinenko <azinenko@amd.com>

rnk committed:
Apply suggestions from code review

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

rnk committed:
Merge remote-tracking branch 'rnk-github/dev-policy-refresh' into dev-policy-refresh

rnk committed:
add significant new text

rnk reviewed (commented):
Thanks for the suggestions, I think it's actually significantly better. Sorry for the long hiatus.

rnk committed:
Apply suggestions from code review

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

rnk commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
-
-Please be aware that all public LLVM mailing lists and discourse forums are public and archived, and
-that notices of confidentiality or non-disclosure cannot be respected.
+and assigned labels. You can subscribe for notification for specific components
```
Yeah, removed.

rnk commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -16,90 +16,87 @@ to eliminate miscommunication, rework, and confusion that might arise from the
+the firehose of all issue notifications, which some community members use to
+perform custom filtering.
+
+Beyond the main formal communication channels, LLVM has a Discord server for
```
I said "asynchronous written", since I think that's the differentiator.

rnk commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -173,10 +171,10 @@ project should be discontinued.
-    * performing these duties for at least three months.
+* engaging in their responsibilities the majority of the days in a month,
+* ensuring that they, and the community members they interact with, abide by the
+  LLVM Community Code of Conduct, and
```
Done.

rnk commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -220,27 +218,41 @@ Test Cases
   :doc:`Testing Guide <TestingGuide>` for details).
 
-* Test cases should be written in :doc:`LLVM assembly language <LangRef>`.
+* We prefer that functional changes are tested using ``FileCheck`` and the tool
```
Done.

rnk committed:
More docs from the review

AaronBallman reviewed (approved):
LGTM aside from a formatting nit. Thank you for working on the refresh!

rnk committed:
Merge branch 'main' into dev-policy-refresh

rnk committed:
Update llvm/docs/DeveloperPolicy.rst

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

rnk committed:
Merge remote-tracking branch 'rnk-github/dev-policy-refresh' into dev-policy-refresh

rnk commented:
Thanks for the review!

I'm about to go on vacation for a while, and given that this is all doc text updates with no intended policy changes, I think I should go ahead and land them. I'm happy to address any unaddressed concerns from @ftynse in two weeks, and I think folks will benefit from the updates in the meantime.
